hydra-main/src/main/java/com/addthis/hydra/job/spawn/balancer/SpawnBalancer.java

```diff
@@ -420,30 +420,32 @@ public boolean canReceiveNewTasks(HostState host) {
                                                          int moveLimit,
                                                          boolean obeyDontAutobalanceMe) {
         List<JobTaskMoveAssignment> rv = purgeMisplacedTasks(host, moveLimit);
-        if (rv.size() <= moveLimit) {
-            long byteLimit = (long) (byteLimitFactor * config.getBytesMovedFullRebalance());
-            List<HostState> hostsSorted = sortHostsByDiskSpace(otherHosts);
-            for (JobTask task : findTasksToMove(host, obeyDontAutobalanceMe)) {
-                long taskTrueSize = getTaskTrueSize(task);
-                if (limitBytes && (taskTrueSize > byteLimit)) {
-                    continue;
-                }
-                JobTaskMoveAssignment assignment = moveTask(task, host.getHostUuid(), hostsSorted);
-                // we don't want to take up one of the limited rebalance slots
-                // with an assignment that we know has no chance of happening
-                // because either the assignment is null or the target host
-                // for the assignment is null
-                if (assignment != null && assignment.getTargetUUID() != null) {
-                    markRecentlyReplicatedTo(assignment.getTargetUUID());
-                    rv.add(assignment);
-                    byteLimit -= taskTrueSize;
-                }
-                if (rv.size() >= moveLimit) {
-                    break;
-                }
+        MoveAssignmentList moveAssignments = new MoveAssignmentList(spawn, taskSizer);
+        moveAssignments.addAll(rv);
+
+        long byteLimit = (long) ((byteLimitFactor * config.getBytesMovedFullRebalance()) - moveAssignments.getBytesUsed());
+        List<HostState> hostsSorted = sortHostsByDiskSpace(otherHosts);
+
+        for (JobTask task : findTasksToMove(host, obeyDontAutobalanceMe)) {
```

Comment:
Yes, perhaps a better way of choosing tasks to move should be used (Kefeng mentioned this as well). I was concerned that moving only the heaviest task from each heavyHost onto the lightHost could make the light host 'heavy' for the next rebalance. This would be less likely to happen with a random selection of tasks. Thoughts on this?